In [1]:
# %matplotlib widget

from __future__ import annotations

import re
from collections import defaultdict
from dataclasses import dataclass, field
from itertools import permutations, product
from math import inf
from random import choice

import matplotlib.colors as mcolors
import matplotlib.pyplot as plt
import networkx as nx
import numpy as np
import numpy.typing as npt
from mpl_toolkits.mplot3d import axes3d
from numpy import int_, object_
from numpy.typing import NDArray
from test_utilities import run_tests_params
from util import print_hex

COLORS = list(mcolors.CSS4_COLORS.keys())

<link href="style.css" rel="stylesheet"></link>
<article class="day-desc read-aloud"><h2>--- Day 12: Digital Plumber ---</h2><p>Walking along the memory banks of the stream, you find a small village that is experiencing a little confusion: some programs can't communicate with each other.</p>
<p>Programs in this village communicate using a fixed system of <em>pipes</em>. Messages are passed between programs using these pipes, but most programs aren't connected to each other directly.  Instead, programs pass messages between each other until the message reaches the intended recipient.</p>
<p>For some reason, though, some of these messages aren't ever reaching their intended recipient, and the programs suspect that some <span title="Yes, citizens, plumbing! It's the latest invention to hit Rome!">pipes</span> are missing. They would like you to investigate.</p>
<p>You walk through the village and record the ID of each program and the IDs with which it can communicate directly (your puzzle input). Each program has one or more programs with which it can communicate, and these pipes are bidirectional; if <code>8</code> says it can communicate with <code>11</code>, then <code>11</code> will say it can communicate with <code>8</code>.</p>
<p>You need to figure out how many programs are in the group that contains program ID <code>0</code>.</p>
<p>For example, suppose you go door-to-door like a travelling salesman and record the following list:</p>
<pre><code>0 &lt;-&gt; 2
1 &lt;-&gt; 1
2 &lt;-&gt; 0, 3, 4
3 &lt;-&gt; 2, 4
4 &lt;-&gt; 2, 3, 6
5 &lt;-&gt; 6
6 &lt;-&gt; 4, 5
</code></pre>
<p>In this example, the following programs are in the group that contains program ID <code>0</code>:</p>
<ul>
<li>Program <code>0</code> by definition.</li>
<li>Program <code>2</code>, directly connected to program <code>0</code>.</li>
<li>Program <code>3</code> via program <code>2</code>.</li>
<li>Program <code>4</code> via program <code>2</code>.</li>
<li>Program <code>5</code> via programs <code>6</code>, then <code>4</code>, then <code>2</code>.</li>
<li>Program <code>6</code> via programs <code>4</code>, then <code>2</code>.</li>
</ul>
<p>Therefore, a total of <code>6</code> programs are in this group; all but program <code>1</code>, which has a pipe that connects it to itself.</p>
<p><em>How many programs</em> are in the group that contains program ID <code>0</code>?</p>
</article>


In [2]:
from collections import deque
from re import findall, split


example = """
0 <-> 2
1 <-> 1
2 <-> 0, 3, 4
3 <-> 2, 4
4 <-> 2, 3, 6
5 <-> 6
6 <-> 4, 5
"""


def program_group_0(lst: str) -> int:
    graph = {}
    for line in lst.strip().splitlines():
        node, *neigbors = map(int, findall(r"\d+", line))
        graph[node] = set(neigbors)

    queue = deque(graph[0])

    while queue:
        node = queue.popleft()
        if node != 0:
            graph[0] |= graph[node]
            queue.extend(graph[node])
            graph[node] = set()

    return len(graph[0])


program_group_0(example)

6

In [3]:
from more_itertools import first


with open("../input/day12.txt") as f:
    puzzle = f.read()


print(f"Part I: {program_group_0(puzzle)}")

Part I: 380


<link href="style.css" rel="stylesheet"></link>
<main>

<p>Your puzzle answer was <code>380</code>.</p><p class="day-success">The first half of this puzzle is complete! It provides one gold star: *</p>
<article class="day-desc"><h2 id="part2">--- Part Two ---</h2><p>There are more programs than just the ones in the group containing program ID <code>0</code>. The rest of them have no way of reaching that group, and still might have no way of reaching each other.</p>
<p>A <em>group</em> is a collection of programs that can all communicate via pipes either directly or indirectly. The programs you identified just a moment ago are all part of the same group. Now, they would like you to determine the total number of groups.</p>
<p>In the example above, there were <code>2</code> groups: one consisting of programs <code>0,2,3,4,5,6</code>, and the other consisting solely of program <code>1</code>.</p>
<p><em>How many groups are there</em> in total?</p>
</article>

</main>


In [8]:
def groups(lst: str) -> int:
    groups = defaultdict(set)
    for line in lst.strip().splitlines():
        node, *neigbors = map(int, findall(r"\d+", line))
        for n in neigbors:
            groups[n].add(node)

    for group in list(groups.keys()):
        queue = deque(groups.get(group, set()))
        while queue:
            child = queue.popleft()
            if child != group:
                queue.extend(groups[child])
                groups[group] |= groups[child]
                del groups[child]

    return len(groups)


print(f"Example = {groups(example)=} should be 2")

Example = groups(example)=2 should be 2


In [5]:
from more_itertools import last


print(f"Part II: {groups(puzzle)}")

Part II: 181


<link href="style.css" rel="stylesheet"></link>
<main>

<p>Your puzzle answer was <code>181</code>.</p><p class="day-success">Both parts of this puzzle are complete! They provide two gold stars: **</p>

</main>


<link href="style.css" rel="stylesheet"></link>


<link href="style.css" rel="stylesheet"></link>
